File: platform/docs/samples/mobile-payment-ios/Sources/MobilePayment/MobilePayment.swift
import Foundation
import LayerXMobile
import LayerXSDK
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

let settledStates: Set<String> = ["done", "done-finalised", "refused"]
let completedStates: Set<String> = ["done", "done-finalised"]

func fail(_ reason: String) -> Never {
    FileHandle.standardError.write(Data("mobile-payment-ios: \(reason)\n".utf8))
    exit(1)
}

func required(_ name: String) -> String {
    guard let value = ProcessInfo.processInfo.environment[name], !value.isEmpty else {
        fail("missing \(name)")
    }
    return value
}

func emit(_ value: [String: JSONValue]) {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    guard let encoded = try? encoder.encode(JSONValue.object(value)) else {
        fail("could not encode the report")
    }
    FileHandle.standardOutput.write(encoded)
    FileHandle.standardOutput.write(Data("\n".utf8))
}

func state(of journey: JSONValue) -> String {
    journey.objectValue?["state"]?.stringValue ?? ""
}

func receipts(in journey: JSONValue) -> [JSONValue] {
    guard case let .array(evidence)? = journey.objectValue?["evidence"] else { return [] }
    return evidence.compactMap { entry in
        guard entry.objectValue?["class"]?.stringValue == "layerx-receipt",
              let identifier = entry.objectValue?["evidence_id"]?.stringValue else { return nil }
        return .string(identifier)
    }
}

@main
struct MobilePayment {
    static func main() async {
        let source = required("LAYERX_SOURCE")
        let destination = required("LAYERX_DESTINATION")
        let money = JSONValue.object([
            "amount": .string(required("LAYERX_AMOUNT")),
            "currency": .string(required("LAYERX_CURRENCY")),
        ])
        guard let paymentKey = IdempotencyKey(rawValue: required("LAYERX_PAYMENT_KEY")) else {
            fail("invalid LAYERX_PAYMENT_KEY")
        }

        do {
            // layerx:begin integration
            let settings = try PublishableConfiguration(environment: ProcessInfo.processInfo.environment)
            let layerx = try LayerXMobile(configuration: settings)
            let quote = try await layerx.client.quote(.object(["source": .string(source), "destination": .string(destination), "money": money]))
            guard let quoteID = quote.objectValue?["quote_id"]?.stringValue else { fail("move quote omitted quote_id") }
            var journey = try await layerx.client.commit(.object(["quote_id": .string(quoteID)]), idempotencyKey: paymentKey)
            // layerx:end integration

            guard let journeyID = journey.objectValue?["journey_id"]?.stringValue else {
                fail("commit omitted journey_id")
            }
            var attempt = 0
            while attempt < 40 && !settledStates.contains(state(of: journey)) {
                try await Task.sleep(nanoseconds: 250_000_000)
                journey = try await layerx.client.journey(id: journeyID)
                attempt += 1
            }

            var report: [String: JSONValue] = [
                "journey_id": .string(journeyID),
                "state": .string(state(of: journey)),
                "receipts": .array(receipts(in: journey)),
                "integration": .string(platform_int_ios().name),
            ]
            if let refusal = journey.objectValue?["refusal"]?.objectValue {
                report["refused_by"] = refusal["refused_by"] ?? .null
                report["money_left"] = refusal["money_left"] ?? .null
            }
            emit(report)
            exit(completedStates.contains(state(of: journey)) ? 0 : 2)
        } catch let error as MobileIntegrationError {
            fail("refused: \(error.code.rawValue)")
        } catch let error as PlatformSDKError {
            fail("refused: \(error.code.rawValue)")
        } catch {
            fail("refused: transport-failure")
        }
    }
}

File: platform/docs/samples/mobile-payment-android/src/main/java/com/sidiora/layerx/docs/MobilePayment.java
package com.sidiora.layerx.docs;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sidiora.layerx.android.LayerXAndroid;
import com.sidiora.layerx.android.PublishableConfiguration;
import com.sidiora.layerx.sdk.IdempotencyKey;
import java.util.List;

public final class MobilePayment {
    private static final List<String> SETTLED = List.of("done", "done-finalised", "refused");
    private static final List<String> COMPLETED = List.of("done", "done-finalised");

    private MobilePayment() {}

    // layerx:begin integration
    static LayerXAndroid openLayerX() {
        return LayerXAndroid.create(PublishableConfiguration.ofEnvironment(System.getenv()));
    }

    static JsonNode pay(LayerXAndroid layerx, ObjectNode move, String paymentKey) {
        var quote = layerx.client().quote(move).toCompletableFuture().join();
        var commit = layerx.mapper().createObjectNode().put("quote_id", quote.path("quote_id").asText());
        return layerx.client().commit(commit, new IdempotencyKey(paymentKey)).toCompletableFuture().join();
    }
    // layerx:end integration

    public static void main(String[] arguments) throws Exception {
        try (LayerXAndroid layerx = openLayerX()) {
            JsonNode journey = settle(layerx, pay(layerx, move(layerx), required("LAYERX_PAYMENT_KEY")));
            System.out.println(report(layerx, journey));
            if (!COMPLETED.contains(state(journey))) {
                System.exit(2);
            }
        }
    }

    private static ObjectNode move(LayerXAndroid layerx) {
        ObjectNode money = layerx.mapper().createObjectNode()
            .put("amount", required("LAYERX_AMOUNT"))
            .put("currency", required("LAYERX_CURRENCY"));
        ObjectNode request = layerx.mapper().createObjectNode()
            .put("source", required("LAYERX_SOURCE"))
            .put("destination", required("LAYERX_DESTINATION"));
        request.set("money", money);
        return request;
    }

    private static JsonNode settle(LayerXAndroid layerx, JsonNode committed) throws InterruptedException {
        JsonNode latest = committed;
        String identifier = latest.path("journey_id").asText();
        for (int attempt = 0; attempt < 40 && !SETTLED.contains(state(latest)); attempt += 1) {
            Thread.sleep(250L);
            latest = layerx.client().journey(identifier).toCompletableFuture().join();
        }
        return latest;
    }

    private static ObjectNode report(LayerXAndroid layerx, JsonNode journey) {
        ObjectNode body = layerx.mapper().createObjectNode();
        body.put("journey_id", journey.path("journey_id").asText());
        body.put("state", state(journey));
        ArrayNode receipts = body.putArray("receipts");
        for (JsonNode reference : journey.path("evidence")) {
            if ("layerx-receipt".equals(reference.path("class").asText())) {
                receipts.add(reference.path("evidence_id").asText());
            }
        }
        if (journey.hasNonNull("refusal")) {
            body.set("refused_by", journey.path("refusal").path("refused_by"));
            body.set("money_left", journey.path("refusal").path("money_left"));
        }
        return body;
    }

    private static String state(JsonNode journey) {
        return journey.path("state").asText();
    }

    private static String required(String name) {
        String value = System.getenv(name);
        if (value == null || value.isEmpty()) {
            throw new IllegalStateException("missing " + name);
        }
        return value;
    }
}
